crates/bevy_render/src/world_sync.rs

```diff
@@ -0,0 +1,252 @@
+use bevy_ecs::{
+    component::Component,
+    entity::Entity,
+    observer::Trigger,
+    query::With,
+    reflect::ReflectComponent,
+    system::{Local, Query, ResMut, Resource, SystemState},
+    world::{Mut, OnAdd, OnRemove, World},
+};
+use bevy_hierarchy::DespawnRecursiveExt;
+use bevy_reflect::Reflect;
+use bevy_utils::tracing::warn;
+
+/// A Plugin that synchronizes entities with specific Components between the main world and render world.
+///
+/// Bevy's renderer is architected independently from main app, It operates in its own separate ECS World. Therefore, the renderer could run in parallel with main app logic, This is called "Pipelined Rendering". See [`PipelinedRenderingPlugin`] for more information.
+///
+/// Previously, `extract` will copy the related main world entity and its data into the render world , and then render world will clear all render entities at the end of frame to reserve enough entity space to ensure that no main world entity ID has been occupied during next `extract`.
+///
+/// With `* as entities`, we should not clear all entities in render world because some core metadata (e.g. [`Component`], [`Query`]) are also stored in the form of entity.
+///
+/// So we turn to an entity-to-entity mapping strategy to sync between main world entity and render world entity, where each `synchronized` main entity has a [`RenderEntity`] component holds an Entity ID pointer to its unique counterpart entity in the render world.
+///
+/// A example for `synchronized` main entity 1v1 and 18v1
+///
+/// ```text
+/// |---------------------------Main World----------------------------|
+/// |  Entity  |                    Component                         |
+/// |-----------------------------------------------------------------|
+/// | ID: 1v1  | PointLight | RenderEntity(ID: 3V1) | SyncRenderWorld |
+/// | ID: 18v1 | PointLight | RenderEntity(ID: 5V1) | SyncRenderWorld |
+/// |-----------------------------------------------------------------|
+///
+/// |----------Render World-----------|
+/// |  Entity  |       Component      |
+/// |---------------------------------|
+/// | ID: 3v1  | MainEntity(ID: 1V1)  |
+/// | ID: 5v1  | MainEntity(ID: 18V1) |
+/// |---------------------------------|
+///
+/// ```
+///
+/// To establish a "Synchronous Relationship", you can add a [`SyncRenderWorld`] component to an entity. Once a `synchronized` main entity is despawned ,its corresponding render entity will be automatically purged in the next `Sync`
+///
+/// Now a single frame of execution looks something like below
+///
+/// ```text
+/// |--------------------------------------------------------------------|
+/// |      |         |          Main   world loop                        |
+/// | Sync | extract |---------------------------------------------------|
+/// |      |         |         Render wrold loop                         |
+/// |--------------------------------------------------------------------|
+/// ```
+///
```

Comment:
Here we could cover the fact that not all entities that are extracted in the render world need to be kept in sync via `SyncRenderWorld` because they are handled differently (without entities) in the render world.